python/fbprophet/diagnostics.py

```diff
@@ -0,0 +1,159 @@
+        weekly_seasonality=model.weekly_seasonality,
+        holidays=model.holidays,
+        seasonality_prior_scale=model.seasonality_prior_scale,
+        changepoint_prior_scale=model.changepoint_prior_scale,
+        holidays_prior_scale=model.holidays_prior_scale,
+        mcmc_samples=model.mcmc_samples,
+        interval_width=model.interval_width,
+        uncertainty_samples=model.uncertainty_samples
+    )
+    return result
+
+
+def _cutoffs(df, horizon, k, period):
+    """Generate cutoff dates
+
+    Parameters
+    ----------
+    df: pd.DataFrame with historical data
+    horizon: pd.Timedelta.
+        Forecast horizon
+    k: Int number.
+        The number of forecasts point.
+    period: pd.Timedelta.
+        Simulated Forecast will be done at every this period.
+
+    Returns
+    -------
+    list of pd.Timestamp
+    """
+    # Allocate memory for result list
+    result = [None] * k
+    # Last cutoff is 'latest date in data - horizon' date
+    cutoff = df['ds'].max() - horizon
+    result[0] = cutoff
+
+    for i in range(1, k):
+        cutoff -= period
+        # If data does not exist in data range (cutoff, cutoff + horizon]
+        if not (((df['ds'] > cutoff) & (df['ds'] <= cutoff + horizon)).any()):
+            # Next cutoff point is 'closest date before cutoff in data - horizon'
+            closest_date = df[df['ds'] <= cutoff].max()['ds']
+            cutoff = closest_date - horizon
+        result[i] = cutoff
+
+    # Sort lines in ascending order
+    return reversed(result)
+
+
+def simulated_historical_forecasts(model, horizon, k, period=None):
+    """Simulated Historical Forecasts.
+        If you would like to know it in detail, read the original paper
+        https://facebookincubator.github.io/prophet/static/prophet_paper_20170113.pdf
+
+    Parameters
+    ----------
+    model: Prophet class object.
+        Fitted Prophet model
+    horizon: string which has pd.Timedelta compatible style.
+        Forecast horizon ('5 days', '3 hours', '10 seconds' etc)
+    k: Int number.
+        The number of forecasts point.
+    period: string which has pd.Timedelta compatible style or None, default None.
+        Simulated Forecast will be done at every this period.
+        0.5 * horizon is used when it is None.
+
+    Returns
+    -------
+    A pd.DataFrame with the forecast, actual value and cutoff.
+    """
+    df = model.history.copy().reset_index(drop=True)
+    horizon = pd.Timedelta(horizon)
+    period = 0.5 * horizon if period is None else pd.Timedelta(period)
+    cutoffs = _cutoffs(df, horizon, k, period)
+    predicts = []
+    for cutoff in cutoffs:
+        # Generate new object with copying fitting options
+        m = _copy(model)
+        # Train model
+        m.fit(df[df['ds'] <= cutoff])
+        # Calculate yhat
+        index_predicted = (df['ds'] > cutoff) & (df['ds'] <= cutoff + horizon)
+        yhat = m.predict(df[index_predicted][['ds']])
```

Comment:
if `model.growth == 'logistic'`, we need to include `df['cap']` here also or it will error.